src/eepromfs.c

```diff
@@ -0,0 +1,888 @@
+    if ( cursor + bytes_to_write >= file->max_cursor )
+    {
+        /* It will, let's shorten it */
+        bytes_to_write = file->max_cursor - cursor;
+    }
+
+    if ( bytes_to_write <= 0 )
+    {
+        /* No bytes to write; bail early */
+        return 0;
+    }
+
+    /* Calculate the starting point to write to EEPROM from */
+    size_t current_block = file->start_block + (cursor / EEPROM_BLOCK_SIZE);
+    size_t block_byte_offset = cursor % EEPROM_BLOCK_SIZE;
+    int bytes_written = 0;
+
+    uint8_t eeprom_buf[EEPROM_BLOCK_SIZE];
+    const uint8_t * ptr = src;
+
+    do {
+        /* Only read in the current block if we need to preserve data from it */
+        if ( bytes_to_write < EEPROM_BLOCK_SIZE || block_byte_offset != 0 )
+        {
+            eeprom_read(current_block, eeprom_buf);
+        }
+        /* Fill the current block with the data to write */
+        while ( bytes_to_write > 0 && block_byte_offset < EEPROM_BLOCK_SIZE )
+        {
+            eeprom_buf[block_byte_offset] = *(ptr++);
+            bytes_written++;
+            bytes_to_write--;
+            block_byte_offset++;
+        }
+        /* Write the block and move on to the next one */
+        eeprom_write(current_block, eeprom_buf);
+        block_byte_offset = 0;
+        current_block++;
+    } while ( bytes_to_write > 0 );
+
+    file->cursor = cursor + bytes_written;
+
+    return bytes_written;
+}
+
+/**
+ * @brief Generates a signature based on the filesystem configuration.
+ * 
+ * The signature is a combination of a "magic value", the
+ * number of files, and the total size of the files in the
+ * filesystem.
+ * 
+ * This all fits into a single EEPROM block (8 bytes).
+ * 
+ * The signature is a compromise between minimizing overhead
+ * of the filesystem and providing a best-effort guarantee
+ * that the data in EEPROM matches the file structure that
+ * the filesystem expects.
+ * 
+ * @return An EEPROM block-sized value containing the signature for the filesystem
+ */
+static const uint64_t eepfs_generate_signature() {
+    /* If each file takes at least one block,
+       at most a 16k EEPROM can hold 255 files.
+       Subtract one file for the signature block. */
+    const uint8_t num_files = eepfs_num_files - 1;
+
+    /* Sum up the total bytes used by all files.
+       At most a 16k EEPROM can hold 2048 bytes. 
+       Skip the first file (the signature block). */
+    uint16_t total_bytes = 0;
+    for ( size_t i = 1; i < eepfs_num_files; ++i )
+    {
+        total_bytes += eepfs_files[i].max_cursor;
+    }
+
+    /* Craft an 8-byte block and return it as a value */
+    uint64_t signature;
+    uint8_t * const sig_bytes = (uint8_t *)&signature;
+    sig_bytes[0] = 'e';
+    sig_bytes[1] = 'e';
+    sig_bytes[2] = 'p';
+    sig_bytes[3] = 'f';
+    sig_bytes[4] = 's';
+    sig_bytes[5] = num_files;
+    sig_bytes[6] = total_bytes >> 8;
+    sig_bytes[7] = total_bytes & 0xFF;
+    return signature;
```

Comment:
This signature is checking for consistent filesystem structure, and I totally see its value.

Another common issue with EEPROM access (not only in N64 world) is that it's basically required to have some sort of consistency checking on actual data being read/written because those things wear out and consistency is never absolutely guaranteed. I think Nintendo was also advising developers to absolutely keep internal checksums in EEPROM data.

This can be done in a followup PR if you want, as it's an additional feature at the end of the day. I'm not sure how it would affect the API though. Some considerations:

 * Reading a 8-byte block of the EEPROM takes ~15ms. This is very slow. Reading a full 128K EEPROM would require almost 4 seconds, so it doesn't sound like validating the content is something that can be done at init time.
 * If each file had a 16bit CRC associated with it (either appended at the start/end of actual data, or in a separate table to be kept in the EEPROM), it would be easy to read/update it when `eepfs_read` / `eepfs_write` are called. That is a good point to do it because the code is already going through all the content bytes.
 * For the stdio filesystem, it's less clear how this would work. Given the stdio API, users might read / write a file only partially, seek through it, etc. so there's not good point where the CRC can be calculated over the whole content. I thought of something quite convoluted/"smart" that might work, but not sure if it's worth the effort:
    * Each file has a 16bit "checksum" associated with it. This can be stored in the header of the filesystem or next to each file, it doesn't matter.
    * This "checksum" is calculated block by block: calculate the CRC16 of each 8 byte block, then xor all the resulting CRC16 together. This means that you can calculate this magic checksum in any order, as long as all blocks of the files are processed.
    * When reading, the cursor API keeps a bitmask of processed blocks. As soon as a new block is fetched, if it's the first time, the CRC16 is calculated and a XOR checksum is updated. Also the block is flagged as "done" in the bitmask. As soon as the bitmask is full (all blocks are processed), all I/O functions start returning error if the checksum doesn't match. As a last ditch, fclose() would read all missing blocks and return error if the checksum doesn't match (though nobody checks fclose... so is it worth it? Maybe yes).
    * When writing, something similar happens. At fclose time, all blocks that were not written are read once, so that their CRC can be calculated and the final checksum can be updated.
    * Maybe checksum checks can be disabled altogether at either mount time, or open time (ab)using a POSIX flag. One option would be to reuse `O_EXCL` (which is mapped to `x` in `fopen` format strings) to signal "I don't want checksum calculation while reading / writing". 
       * if the checkum is "appended" at the end of each file, one option would be for `O_EXCL` to switch to "raw mode" where the checksum is returned as part of the actual file content, and performing zero calculations on it. This would also to document to always use `O_EXCL` if the game code wants to perform its own checksum on data and completely bypass the filesystem support.

Reply:
This is an interesting problem that I'll definitely give some more thought and research into.

I considered doing some sort of checksum for each file (or each block) but ultimately punted on the idea due to the new realities of EEPROM save reliability in the age of emulators and flash carts. The vast majority of people running games/software built by libdragon are likely not actually using a cartridge with real EEPROM. There are some programmable cartridges out there (the [N64 Blaster by RetroStage](https://retrostage.net/index.php/product/n64-blaster-2-0/), and the [N64 Project by RetroCircuits](https://www.mortoffgames.com/index.php?route=product/product&path=294&product_id=877)) which support EEPROM saves, but it's unclear whether they actually use EEPROM (further investigation needed).

I think it's a worthwhile follow-up to explore how best to effectively implement checksums with minimal overhead. Hopefully it will be possible to drop-in without much (or any) of a noticeable change on the API side.